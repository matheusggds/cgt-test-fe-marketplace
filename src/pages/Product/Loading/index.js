import { Skeleton } from "@mui/material";
import React from "react";

const Loading = () => (
  <div data-testid="product-page-loading">
    <Skeleton variant="rectangular" width="100%" height={150} sx={{ mb: 2 }} />
    <Skeleton variant="rectangular" height={20} sx={{ mb: 1 }} />
    <Skeleton variant="rectangular" height={20} sx={{ mb: 1 }} />
    <Skeleton variant="rectangular" height={20} sx={{ mb: 1 }} />
    <Skeleton variant="rectangular" height={20} sx={{ mb: 1 }} />
  </div>
);

export default Loading;
