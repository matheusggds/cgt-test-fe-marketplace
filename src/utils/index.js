// convert to dolar string
export const formatToCurrency = (val) =>
  Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
  }).format(val);
