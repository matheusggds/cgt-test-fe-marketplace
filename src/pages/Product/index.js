import React, { useEffect, useState } from "react";

import axios from "axios";
import {
  Alert,
  Box,
  Button,
  Chip,
  Divider,
  Grid,
  Typography,
} from "@mui/material";
import { useParams } from "react-router-dom";
import { formatToCurrency } from "../../utils";

import { Stack } from "@mui/system";
import SelectQuantity from "../../components/SelectQuantity";
import useCartContext from "../../context/cart";
import Error from "../../components/Error";
import Loading from "./Loading";

const STATUS = {
  SUCCESS: "success",
  LOADING: "loading",
  ERROR: "error",
  ADDED_TO_CART: "added_to_cart",
};

const Home = () => {
  const { id } = useParams();
  const [productData, setProductData] = useState();
  const [status, setStatus] = useState(STATUS.LOADING);
  const [quantityToAdd, setQuantityToAdd] = useState(0);
  const [, { addItem }] = useCartContext();

  useEffect(() => {
    const getProducts = async () => {
      try {
        const { data } = await axios.get(
          "https://dummyjson.com/products/" + id
        );

        setProductData(data);
        setStatus(STATUS.SUCCESS);
      } catch (e) {
        setStatus(STATUS.ERROR);
      }
    };

    getProducts();
  }, [id]);

  const AddToCart = () => {
    addItem(productData, quantityToAdd, null, () => {
      setStatus(STATUS.ADDED_TO_CART);

      setTimeout(() => {
        setStatus(STATUS.SUCCESS);
      }, 3000);
    });
  };

  if (status === STATUS.LOADING) {
    return <Loading />;
  }

  if (status === STATUS.ERROR) {
    return <Error />;
  }

  return (
    <>
      <Typography variant="h3" component="h1" gutterBottom>
        {productData.title}
      </Typography>
      <Divider sx={{ marginBottom: "20px" }} />
      <Grid container spacing={2}>
        <Grid item sm={6}>
          <Box mb={3}>
            <img
              src={productData.images[0]}
              title={productData.title}
              alt={productData.title}
            />
          </Box>
        </Grid>
        <Grid item sm={6}>
          <div>
            <Stack direction="row" spacing={1} marginBottom={3}>
              <Chip
                label={productData.brand}
                variant="outlined"
                color="primary"
              />
              <Chip
                label={productData.category}
                variant="outlined"
                color="primary"
              />
            </Stack>
          </div>
          <div>
            <Typography paragraph>{productData.description}</Typography>
          </div>
          <Typography variant="h4">
            {formatToCurrency(productData.price)}
          </Typography>
          <Box sx={{ marginTop: "50px" }}>
            <Stack direction="row" spacing={3}>
              <SelectQuantity
                elId={id}
                quantity={quantityToAdd}
                getValueChanged={(value) => setQuantityToAdd(value)}
              />
              <Button
                color="primary"
                variant="contained"
                fullWidth
                onClick={() => AddToCart()}
              >
                add to cart
              </Button>
            </Stack>
          </Box>
          <Box sx={{ mt: 3 }}>
            {status === STATUS.ADDED_TO_CART && (
              <Alert severity="success">Added to cart</Alert>
            )}
          </Box>
        </Grid>
      </Grid>
    </>
  );
};

export default Home;
