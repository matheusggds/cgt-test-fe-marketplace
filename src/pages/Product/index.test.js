import {
  fireEvent,
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
  within,
} from "@testing-library/react";
import { renderWithContext } from "../../utils";

import axios from "axios";

jest.mock("axios");

import Product from "./";
import { wait } from "@testing-library/user-event/dist/utils";

const MOCK = {
  id: 1,
  title: "title A",
  description: "description A",
  images: ["src/path/1.jpg"],
  price: 5,
};

describe("<Product />", () => {
  describe("View", () => {
    it("should display default data", async () => {
      axios.get.mockReturnValue(Promise.resolve({ data: MOCK }));

      const products = renderWithContext(<Product />);

      await screen.findByText("title A");

      expect(products).toMatchSnapshot();
    });

    it("should display loading screen", async () => {
      axios.get.mockReturnValue(Promise.resolve({ data: MOCK }));

      renderWithContext(<Product />);

      await waitFor(() => {
        expect(screen.getByTestId("product-page-loading")).toBeInTheDocument();
      });
    });

    it("should display error screen", async () => {
      axios.get.mockReturnValue(Promise.reject());

      renderWithContext(<Product />);

      await waitFor(() => {
        expect(
          screen.getByText("Oops! We have a error, try again later")
        ).toBeInTheDocument();
      });
    });
  });

  describe("Requests", () => {
    it("should call API to get products", async () => {
      axios.get.mockReturnValue(Promise.resolve({ data: MOCK }));

      renderWithContext(<Product />);

      waitFor(() => {
        expect(axios.get).toHaveBeenCalled();
        expect(axios.get).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe("Actions", () => {
    it("should add to cart", async () => {
      axios.get.mockReturnValue(Promise.resolve({ data: MOCK }));

      renderWithContext(<Product />);

      await screen.findByText("title A");

      const buttons = screen.getAllByRole("button");

      fireEvent.mouseDown(buttons[0]);

      fireEvent.click(
        screen.getByText(8),
        new MouseEvent("click", {
          bubbles: true,
          cancelable: true,
        })
      );

      fireEvent.click(
        buttons[1],
        new MouseEvent("click", {
          bubbles: true,
          cancelable: true,
        })
      );

      await waitFor(() => {
        expect(screen.getByText("Added to cart")).toBeInTheDocument();
      });

      await waitForElementToBeRemoved(screen.queryByText("Added to cart"), {
        timeout: 4000,
      });
    });
  });
});
