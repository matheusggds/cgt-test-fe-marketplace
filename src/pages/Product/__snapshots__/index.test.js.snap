// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> View should display default data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom css-1mktp8m-MuiTypography-root"
      >
        title A
      </h1>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-gsm6zr-MuiDivider-root"
      />
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6 css-6dzwqk-MuiGrid-root"
        >
          <div
            class="MuiBox-root css-i3pbo"
          >
            <img
              alt="title A"
              src="src/path/1.jpg"
              title="title A"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6 css-6dzwqk-MuiGrid-root"
        >
          <div>
            <div
              class="MuiStack-root css-85gxt3-MuiStack-root"
            >
              <div
                class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-outlinedPrimary css-1t3gu6w-MuiChip-root"
              >
                <span
                  class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                />
              </div>
              <div
                class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-outlinedPrimary css-1t3gu6w-MuiChip-root"
              >
                <span
                  class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                />
              </div>
            </div>
          </div>
          <div>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph css-z2eky3-MuiTypography-root"
            >
              description A
            </p>
          </div>
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            $5.00
          </h4>
          <div
            class="MuiBox-root css-jn01mi"
          >
            <div
              class="MuiStack-root css-u10pvi-MuiStack-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-sizeSmall  css-1amlwio-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="cart-select-quantity cart-select-quantity"
                  class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="cart-select-quantity"
                  role="button"
                  tabindex="0"
                >
                  0
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  tabindex="-1"
                  value="0"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-hdw1oc"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                add to cart
              </button>
            </div>
          </div>
          <div
            class="MuiBox-root css-178yklu"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <h1
      class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom css-1mktp8m-MuiTypography-root"
    >
      title A
    </h1>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-gsm6zr-MuiDivider-root"
    />
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6 css-6dzwqk-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-i3pbo"
        >
          <img
            alt="title A"
            src="src/path/1.jpg"
            title="title A"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6 css-6dzwqk-MuiGrid-root"
      >
        <div>
          <div
            class="MuiStack-root css-85gxt3-MuiStack-root"
          >
            <div
              class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-outlinedPrimary css-1t3gu6w-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
              />
            </div>
            <div
              class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-outlinedPrimary css-1t3gu6w-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
              />
            </div>
          </div>
        </div>
        <div>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph css-z2eky3-MuiTypography-root"
          >
            description A
          </p>
        </div>
        <h4
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          $5.00
        </h4>
        <div
          class="MuiBox-root css-jn01mi"
        >
          <div
            class="MuiStack-root css-u10pvi-MuiStack-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-sizeSmall  css-1amlwio-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="cart-select-quantity cart-select-quantity"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="cart-select-quantity"
                role="button"
                tabindex="0"
              >
                0
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="0"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-hdw1oc"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              add to cart
            </button>
          </div>
        </div>
        <div
          class="MuiBox-root css-178yklu"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
